File: server/services/matchRoomService.js
const MatchRoom = require("../models/MatchRoom");
const Tournament = require("../models/Tournament");

const ApiError = require("../utils/ApiError");
const { ROLES } = require("../constants/roles");

const isPlatformAdmin = (user) => {
  return user.role === ROLES.ADMIN || user.role === ROLES.SUPER_ADMIN;
};

const isTournamentOwner = (tournament, user) => {
  if (!tournament.createdBy) return false;

  return tournament.createdBy.toString() === user._id.toString();
};

const ensureCanManageTournament = (tournament, currentUser) => {
  if (isPlatformAdmin(currentUser)) return;

  if (
    currentUser.role === ROLES.ORGANIZER &&
    isTournamentOwner(tournament, currentUser)
  ) {
    return;
  }

  throw new ApiError(
    403,
    "You can manage match rooms only for your own tournaments",
  );
};

const populateMatchRoom = (query) => {
  return query
    .populate("tournament", "title game mode status createdBy")
    .populate("createdBy", "name email role");
};

const createMatchRoom = async (tournamentId, roomData, currentUser) => {
  const { matchNumber, map, roomId, roomPassword, matchTime } = roomData;

  const tournament = await Tournament.findById(tournamentId);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  ensureCanManageTournament(tournament, currentUser);

  if (tournament.status === "completed") {
    throw new ApiError(
      400,
      "Cannot create match room for completed tournament",
    );
  }

  const existingRoomId = await MatchRoom.findOne({
    roomId,
  });

  if (existingRoomId) {
    throw new ApiError(400, "Room ID already exists");
  }

  const existingMatchNumber = await MatchRoom.findOne({
    tournament: tournamentId,
    matchNumber,
  });

  if (existingMatchNumber) {
    throw new ApiError(400, "Match number already exists for this tournament");
  }

  const matchRoom = await MatchRoom.create({
    tournament: tournamentId,
    matchNumber,
    map,
    roomId,
    roomPassword,
    matchTime,
    createdBy: currentUser._id,
  });

  return populateMatchRoom(MatchRoom.findById(matchRoom._id));
};
const getAllMatchRooms = async () => {
  const matchRooms = await populateMatchRoom(
    MatchRoom.find().select("-roomPassword").sort({ createdAt: -1 }),
  );

  return matchRooms;
};

const getMyCreatedMatchRooms = async (currentUser) => {
  if (isPlatformAdmin(currentUser)) {
    const matchRooms = await populateMatchRoom(
      MatchRoom.find().sort({ createdAt: -1 }),
    );

    return matchRooms;
  }

  const matchRooms = await populateMatchRoom(
    MatchRoom.find({
      createdBy: currentUser._id,
    }).sort({ createdAt: -1 }),
  );

  return matchRooms;
};

const getMatchRoomById = async (matchRoomId) => {
  const matchRoom = await populateMatchRoom(MatchRoom.findById(matchRoomId));

  if (!matchRoom) {
    throw new ApiError(404, "Match room not found");
  }

  return matchRoom;
};

const updateMatchRoom = async (matchRoomId, updateData, currentUser) => {
  const matchRoom = await MatchRoom.findById(matchRoomId);

  if (!matchRoom) {
    throw new ApiError(404, "Match room not found");
  }

  const tournament = await Tournament.findById(matchRoom.tournament);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  ensureCanManageTournament(tournament, currentUser);

  if (tournament.status === "completed") {
    throw new ApiError(
      400,
      "Cannot update match room for completed tournament",
    );
  }

  if (updateData.roomId !== undefined) {
    const existingRoom = await MatchRoom.findOne({
      roomId: updateData.roomId,
      _id: { $ne: matchRoomId },
    });

    if (existingRoom) {
      throw new ApiError(400, "Room ID already exists");
    }

    matchRoom.roomId = updateData.roomId;
  }

  if (updateData.matchNumber !== undefined) {
    const existingMatchNumber = await MatchRoom.findOne({
      tournament: matchRoom.tournament,
      matchNumber: updateData.matchNumber,
      _id: { $ne: matchRoomId },
    });

    if (existingMatchNumber) {
      throw new ApiError(
        400,
        "Match number already exists for this tournament",
      );
    }

    matchRoom.matchNumber = updateData.matchNumber;
  }

  if (updateData.map !== undefined) {
    matchRoom.map = updateData.map;
  }

  if (updateData.roomPassword !== undefined) {
    matchRoom.roomPassword = updateData.roomPassword;
  }

  if (updateData.matchTime !== undefined) {
    matchRoom.matchTime = updateData.matchTime;
  }

  await matchRoom.save();

  return populateMatchRoom(MatchRoom.findById(matchRoom._id));
};

const deleteMatchRoom = async (matchRoomId, currentUser) => {
  const matchRoom = await MatchRoom.findById(matchRoomId);

  if (!matchRoom) {
    throw new ApiError(404, "Match room not found");
  }

  const tournament = await Tournament.findById(matchRoom.tournament);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  ensureCanManageTournament(tournament, currentUser);

  if (tournament.status === "completed") {
    throw new ApiError(
      400,
      "Cannot delete match room for completed tournament",
    );
  }

  await matchRoom.deleteOne();

  return matchRoom;
};

module.exports = {
  createMatchRoom,
  getAllMatchRooms,
  getMyCreatedMatchRooms,
  getMatchRoomById,
  updateMatchRoom,
  deleteMatchRoom,
};

File: backend/app/services/match_room_service.py
from fastapi import HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session
from app.models.domain import MatchRoom,Tournament,Team,TeamMember,TournamentRegistration
from app.constants.roles import Role

def can_manage(t,u): return u.role in {Role.ADMIN,Role.SUPER_ADMIN} or (u.role==Role.ORGANIZER and t.created_by==u.id)
def create(db,u,tid,d):
    t=db.get(Tournament,tid)
    if not t: raise HTTPException(404,"Tournament not found")
    if not can_manage(t,u): raise HTTPException(403,"You can manage match rooms only for your own tournaments")
    if t.status.value=='completed': raise HTTPException(400,"Cannot create match room for completed tournament")
    if db.scalar(select(MatchRoom).where(MatchRoom.room_id==d.roomId)): raise HTTPException(400,"Room ID already exists")
    if db.scalar(select(MatchRoom).where(MatchRoom.tournament_id==tid,MatchRoom.match_number==d.matchNumber)): raise HTTPException(400,"Match number already exists for this tournament")
    r=MatchRoom(room_id=d.roomId,room_password=d.roomPassword,match_number=d.matchNumber,map=d.map,match_time=d.matchTime,tournament_id=tid,created_by=u.id); db.add(r); db.commit(); db.refresh(r); return r
def all_rooms(db): return db.scalars(select(MatchRoom).order_by(MatchRoom.created_at.desc())).unique().all()
def my_created(db,u):
    q=select(MatchRoom).order_by(MatchRoom.created_at.desc())
    if u.role not in {Role.ADMIN,Role.SUPER_ADMIN}: q=q.where(MatchRoom.created_by==u.id)
    return db.scalars(q).unique().all()
def eligible(db,u):
    team=db.scalar(select(Team).join(TeamMember).where(TeamMember.user_id==u.id))
    if not team: return []
    tids=db.scalars(select(TournamentRegistration.tournament_id).where(TournamentRegistration.team_id==team.id)).all()
    if not tids:return []
    return db.scalars(select(MatchRoom).where(MatchRoom.tournament_id.in_(tids)).order_by(MatchRoom.match_time)).unique().all()
def get(db,rid):
    r=db.get(MatchRoom,rid)
    if not r: raise HTTPException(404,"Match room not found")
    return r
def update(db,u,rid,d):
    r=get(db,rid); t=r.tournament
    if not can_manage(t,u): raise HTTPException(403,"You can manage match rooms only for your own tournaments")
    if t.status.value=='completed': raise HTTPException(400,"Cannot update match room for completed tournament")
    if d.roomId is not None and db.scalar(select(MatchRoom).where(MatchRoom.room_id==d.roomId,MatchRoom.id!=r.id)): raise HTTPException(400,"Room ID already exists")
    if d.matchNumber is not None and db.scalar(select(MatchRoom).where(MatchRoom.tournament_id==r.tournament_id,MatchRoom.match_number==d.matchNumber,MatchRoom.id!=r.id)): raise HTTPException(400,"Match number already exists for this tournament")
    for k,v in [("room_id",d.roomId),("room_password",d.roomPassword),("match_number",d.matchNumber),("map",d.map),("match_time",d.matchTime)]:
        if v is not None:setattr(r,k,v)
    db.commit(); db.refresh(r); return r
def delete(db,u,rid):
    r=get(db,rid)
    if not can_manage(r.tournament,u): raise HTTPException(403,"You can manage match rooms only for your own tournaments")
    if r.tournament.status.value=='completed': raise HTTPException(400,"Cannot delete match room for completed tournament")
    db.delete(r); db.commit()
